File: 3776.minimum-moves-to-balance-circular-array.cpp
#
# @lc app=leetcode id=3776 lang=cpp
#
# [3776] Minimum Moves to Balance Circular Array
#
# @lc code=start
class Solution {
public:
    long long minMoves(vector<int>& balance) {
        int n = balance.size();
        long long total = 0;
        
        // Check if solution is possible
        for (int x : balance) {
            total += x;
        }
        
        if (total < 0) return -1;
        
        long long minCost = LLONG_MAX;
        
        // Try all possible starting positions to linearize the circular array
        for (int start = 0; start < n; start++) {
            long long cost = 0;
            long long carry = 0;
            
            // Calculate cost for this linearization
            for (int i = 0; i < n - 1; i++) {
                carry += balance[(start + i) % n];
                cost += abs(carry);
            }
            
            minCost = min(minCost, cost);
        }
        
        return minCost;
    }
};
# @lc code=end

File: 3776.minimum-moves-to-balance-circular-array.py
#
# @lc app=leetcode id=3776 lang=python3
#
# [3776] Minimum Moves to Balance Circular Array
#

# @lc code=start
from typing import List

class Solution:
    def minMoves(self, balance: List[int]) -> int:
        n = len(balance)
        total = sum(balance)
        if total < 0:
            return -1
        if all(x >= 0 for x in balance):
            return 0
        # Find the index with the negative balance
        deficit_idx = -1
        for i, val in enumerate(balance):
            if val < 0:
                deficit_idx = i
                break
        deficit = -balance[deficit_idx]
        # Collect all surplus positions and their values
        surplus = []
        for i, val in enumerate(balance):
            if val > 0:
                # Distance in circular array
                dist = min((i - deficit_idx) % n, (deficit_idx - i) % n)
                surplus.append((dist, i, val))
        surplus.sort()  # process closest first
        moves = 0
        for dist, idx, val in surplus:
            if deficit <= 0:
                break
            take = min(deficit, val)
            moves += take * dist
            deficit -= take
        # Final check: all balances should be non-negative
        if deficit > 0:
            return -1
        return moves
# @lc code=end